docs/tooling-consolidation.md

```diff
@@ -0,0 +1,82 @@
+# AspNetCore Tooling Consolidation
+
+## Objectives
+
+We want to consolidate portions of dotnet/aspnetcore-tooling into dotnet/aspnetcore in our ongoing effort to reduce overall build time end to end for the .NET Core SDK. As such, we will not require all of aspnetcore-tooling to be merged into aspnetcore. For example, tooling for VSMac and VSCode will remain in aspnetcore-tooling. The main motivation is to achieve the overall goals of repo consolidation to reduce the number of repository required to build the SDK while keeping the build in aspnetcore as simple as possible without needing to support build/test for tooling scenarios such as testing on VSCode. There is also an added benefit of maintaining the current development and release workflow for aspnetcore-tooling which is more "agile" than aspnetcore (e.g. faster PR builds, release cycles that synchronize with VS releases).
+
+To achieve this we will be migrating the following (and associated tests) from aspnetcore-tooling to aspnetcore:
+```
+Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X
+Microsoft.AspNetCore.Mvc.Razor.Extensions.Version2_X
+Microsoft.AspNetCore.Mvc.Razor.Extensions
+Microsoft.AspNetCore.Razor.Language
+Microsoft.AspNetCore.Razor.Tools
+Microsoft.CodeAnalysis.Razor
+Microsoft.NET.Sdk.Razor
+````
+The following (and associated tests) will remain in aspnetcore-tooling:
+```
+Microsoft.AspNetCore.Razor.LanguageServer.Common
+Microsoft.AspNetCore.Razor.LanguageServer
+Microsoft.AspNetCore.Razor.OmniSharpPlugin.StrongNamed
+Microsoft.AspNetCore.Razor.OmniSharpPlugin
+Microsoft.AspNetCore.Razor.VSCode.Extension
+Microsoft.AspNetCore.Razor.VSCode
+Microsoft.CodeAnalysis.Razor.Workspaces
+Microsoft.CodeAnalysis.Remote.Razor
+Microsoft.VisualStudio.Editor.Razor
+Microsoft.VisualStudio.LanguageServerClient.Razor
+Microsoft.VisualStudio.LanguageServices.Razor
+Microsoft.VisualStudio.LiveShare.Razor
+Microsoft.VisualStudio.Mac.LanguageServices.Razor
+Microsoft.VisualStudio.Mac.RazorAddin
+Microsoft.VisualStudio.RazorExtension
+RazorDeveloperTools
+rzls
+````
+
+Due to the separation of tooling projects, the new dependency flow will be:
+
+```
+runtime +--> aspnetcore +--> SDK
+        \               \
+         +-> extensions -+-> aspnetcore-tooling
+```
+## Known action items
+
+1. Migrate source with `git filter-branch`, preserving commit history
+2. Migration of pending PRs to aspnetcore after source migration.
+   1. Alternatively, we can freeze checkins during source migration especially since there isn't much source code that requires migration
+3. Build razor SDK in pre-build step along with RepoTools and install it locally before the main build in aspnetcore
+   1. Can this step be skipped if for devs who don't work in this area?
```

Comment:
We made the assumption going in that the Razor SDK needs to be prebuilt but is that actually the case? For example, could the Microsoft.NET.Sdk.Razor project do whatever installation is needed when it's built and can the project do _nothing_ if the installation location is up-to-date?

If prebuilding is required, spell out why and include information about how long it will take, how late in the normal `./build` process projects that need the SDK build, et cetera

Reply:
I would assume it's so that we can build our projects with the razor sdk like so:
https://github.com/dotnet/aspnetcore/blob/3787d7e7f070543cc9368d589a504fa8c4bd4830/src/Azure/AzureAD/Authentication.AzureADB2C.UI/src/Microsoft.AspNetCore.Authentication.AzureADB2C.UI.csproj#L1
https://github.com/dotnet/aspnetcore/blob/3787d7e7f070543cc9368d589a504fa8c4bd4830/src/Components/Authorization/src/Microsoft.AspNetCore.Components.Authorization.csproj#L1
@NTaylorMullen is that correct?

Reply:
I believe so but @pranavkm would know better

Reply:
This should be solveable without having to pre-build the SDK. These projects already reference the SDK as a package: https://github.com/dotnet/aspnetcore/blob/20139447c148ec89cc211e740804c67540fe77a6/eng/targets/CSharp.Common.props#L17-L23, we would need to replace this with a project reference + directly importing the props file \ targets file from the SDK directly.

The one issue with referencing the SDK is that it locks up tasks dll that the SDK produces. We run in to the same with WebAssembly.Build, but have a tolerable solution to it: https://github.com/dotnet/aspnetcore/blob/blazor-wasm/src/Components/WebAssembly/Build/src/Microsoft.AspNetCore.Components.WebAssembly.Build.csproj#L47-L71

Reply:
Interesting, if we're able to reference SDKs in that way, then we do not need to separate it into a different build stage. I'll update the docs to reflect this as the approach we are going to take. I'll mention the prebuild as a "fall back" plan in case we find difficulties but I don' anticipate we'll need to do that.